### PR TITLE
feat: usage of predict_linear function in volume dashboard

### PR DESCRIPTION
**Pull request by Hardikl**
image

### CODE REVIEW

File: grafana/dashboards/cmode/volume.json [hunk 1]
```diff
@@ -7131,6 +7131,267 @@
...
+              "interval": "",
+              "legendFormat": "Actual - {{svm}} - {{volume}}",
+              "range": true,
+              "refId": "B"
+            },
+            {
+              "datasource": "${DS_PROMETHEUS}",
+              "disableTextWrap": false,
+              "editorMode": "code",
+              "expr": "predict_linear(volume_space_physical_used{datacenter=~\"$Datacenter\",cluster=~\"$Cluster\",svm=~\"$SVM\",volume=~\"$Volume\"}[$__range], 30*24*3600) and topk($TopResources, avg_over_time(volume_space_physical_used{datacenter=~\"$Datacenter\",cluster=~\"$Cluster\",svm=~\"$SVM\",volume=~\"$Volume\"}[$__range] @ end()))",
```
[the remark this answers is not in the document]
Comment (Hardikl) [reply grafana/dashboards/cmode/volume.json:7250]: Here, 3 things to be considered.
**Query:**
predict_linear(volume_xxxxxx{}[time_range], predict_time)
**Timeshift:**
"timeShift": "0d+30d",

1. **time_range**: This is past time range which would fed to predict linear function as input -> mostly same as dashboard time range. Used in Prometheus
2. **predict_time**: This is future time range which predict linear function predict as output. Used in Prometheus.
3. **timeshift**: This is time which shifted left/right(past/future) to visualise panel better. Used in Grafana.

Relation between them are as below
- time_range >= predict_time, (ideal case, time_range = 2 x predict_time)
- predict_time == timeshift

[the remark this answers is not in the document]
Comment (Hardikl) [reply grafana/dashboards/cmode/volume.json:7250]: image

**Comment by Hardikl** [comment]
Update from the investigation so far,

As discussed, I have tried to use the `predict_linear` function this way, Sample interval: 3 days, Prediction interval: 15 days.

There are few challenges while plotting this graph in Grafana
- Query seem to be too slower as `predict linear` function takes `topk` query as input which is also slower in nature due to range and `avg_over_time`.
- While plotting the graph, If we use sample interval as `3d` then we can't provide the `Time shift` for more than 3d. 
`Time shift` is the value which can shift whole graph in right(future) or left(past) direction. 

My idea was to plot a graph this way:
from current_date-15d(past) to current_date+15d(future) and current date will be at center. In this case, `Time shift` should be 15d.

**Comment by Hardikl** [comment]
image

**Comment by Hardikl** [comment]
image

File: grafana/dashboards/cmode/volume.json [hunk 2]
```diff
@@ -7573,6 +7573,397 @@
...
+                  "Value #C": false,
+                  "__name__": true,
+                  "cluster": true,
+                  "datacenter": true,
+                  "instance": true,
+                  "job": true,
+                  "node": true,
+                  "state": true,
+                  "style": true,
+                  "{__name__=\"volume_space_physical_used\", aggr=\"harvest_vc_aggr\", cluster=\"umeng-aff300-05-06\", datacenter=\"rest\", instance=\"localhost:13003\", job=\"c3\", node=\"umeng-aff300-05\", style=\"flexgroup\", svm=\"umeng-aff300-05-svm1\", volume=\"umeng_harvest_vol1\"}": true
```
[the remark this answers is not in the document]
Comment (Hardikl) [reply grafana/dashboards/cmode/volume.json:7729]: removed

**Comment by Hardikl** [comment]
Instead of using `physical used` and `logical used`, now using `size used` with clamping at 0% and 100% and made table size as full width which shows node and cluster data as well.
image

File: grafana/dashboards/cmode/volume.json [hunk 3]
```diff
@@ -7573,6 +7573,397 @@
...
+                "displayName": "space used"
+              }
+            ]
+          },
+          "pluginVersion": "10.3.1",
+          "targets": [
+            {
+              "datasource": "${DS_PROMETHEUS}",
+              "editorMode": "code",
+              "expr": "predict_linear(volume_space_physical_used_percent{datacenter=~\"$Datacenter\",cluster=~\"$Cluster\",svm=~\"$SVM\",volume=~\"$Volume\"}[3d], 15*24*3600) and topk($TopResources, avg_over_time(volume_space_physical_used_percent{datacenter=~\"$Datacenter\",cluster=~\"$Cluster\",svm=~\"$SVM\",volume=~\"$Volume\"}[$__range] @ end()))",
```
[the remark this answers is not in the document]
Comment (Hardikl) [reply grafana/dashboards/cmode/volume.json]: Using `size_used_percent` metric which always <= 100.

File: grafana/dashboards/cmode/volume.json [hunk 4]
```diff
@@ -7561,6 +7561,282 @@
       ],
       "title": "Volume Growth Rate",
       "type": "row"
+    },
+    {
```
[the remark this answers is not in the document]
Comment (Hardikl) [reply grafana/dashboards/cmode/volume.json:7565]: - Datacenter column added
- Cluster column moved to 2nd
- Node column removed
- % changed to gradient 
- Default sort desc for predicted value.


image

**Review by cgrinds — APPROVED** [review]

**Pull request merged**